internal/controller/nonadmindownloadrequest_controller.go

```diff
@@ -0,0 +1,292 @@
+	}
+	// TODO: delete if UID is always available here.
+	// if req.ObjectMeta.UID  == "" {
+	// requeue later until uid is populated
+	// }
+	logger := log.FromContext(ctx)
+	logger.Info("Reconciling NonAdminDownloadRequest", "name", req.Name, "namespace", req.Namespace)
+	// defines associated downloadrequest for getting, or deleting
+	veleroDR := velerov1.DownloadRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.VeleroDownloadRequestName(),
+			Namespace: r.OADPNamespace,
+			Labels: func() map[string]string {
+				nal := function.GetNonAdminLabels()
+				nal[constant.NadrOriginNACUUIDLabel] = string(req.GetUID())
+				return nal
+			}(),
+			Annotations: function.GetNonAdminDownloadRequestAnnotations(req),
+		},
+		Spec: velerov1.DownloadRequestSpec{
+			Target: velerov1.DownloadTarget{
+				Kind: req.Spec.Target.Kind,
+			},
+		},
+	}
+	// if request has status, it may already be processed or expired
+	if req.Status.VeleroDownloadRequest.Status != nil {
+		// if request is expired, delete NADR after deleting velero DR
+		if req.Status.VeleroDownloadRequest.Status.Expiration != nil && req.Status.VeleroDownloadRequest.Status.Expiration.Before(&metav1.Time{Time: time.Now()}) {
+			// find associated velero downloadrequest and delete that first
+			logger.V(1).Info("Deleting expired NonAdminDownloadRequest associated velero download request", req.VeleroDownloadRequestName(), req.Namespace)
+			if err := r.Delete(ctx, &veleroDR); err != nil {
+				if !apierrors.IsNotFound(err) {
+					logger.V(1).Info("Failed to delete expired NonAdminDownloadRequest associated velero download request", req.VeleroDownloadRequestName(), err)
+					// other errors, requeue to retry delete
+					return reconcile.Result{}, err
+				}
+			}
+			logger.V(1).Info("Deleting expired NonAdminDownloadRequest", req.Name, req.Namespace)
+			if err := r.Delete(ctx, req); err != nil {
+				if !apierrors.IsNotFound(err) {
+					logger.V(1).Info("Failed to delete expired NonAdminDownloadRequest", req.Name, err)
+					// other errors, requeue to retry delete
+					return reconcile.Result{}, err
+				}
+				// not found, stop requeue
+				return reconcile.Result{Requeue: false}, nil
+			}
+		}
+		// if request is not expired, and has downloadUrl, requeue when expired to cleanup.
+		if req.Status.VeleroDownloadRequest.Status.DownloadURL != constant.EmptyString {
+			return reconcile.Result{RequeueAfter: time.Until(req.Status.VeleroDownloadRequest.Status.Expiration.Time)}, nil
+		}
+	}
+	// try get veleroDR if exists, then update status
+	if err := r.Get(ctx, types.NamespacedName{Namespace: veleroDR.Namespace, Name: veleroDR.Name}, &veleroDR); err == nil {
+		// wait for next reconcile?
+		// if url is available and not expired, then set status to completed
+		if veleroDR.Status.DownloadURL != constant.EmptyString &&
+			veleroDR.Status.Phase == velerov1.DownloadRequestPhaseProcessed &&
+			!veleroDR.Status.Expiration.Before(&metav1.Time{Time: time.Now()}) {
```

Comment:
There is also `realClock.Now()` which is more test friendly ?

Reply:
checking